extractor/filesystem/language/javascript/pnpmlock/extractor.go

```diff
@@ -0,0 +1,274 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package pnpmlock extracts pnpm-lock.yaml files.
+package pnpmlock
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/go-yaml/yaml"
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem"
+	"github.com/google/osv-scalibr/extractor/filesystem/osv"
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+)
+
+type pnpmLockPackageResolution struct {
+	Tarball string `yaml:"tarball"`
+	Commit  string `yaml:"commit"`
+	Repo    string `yaml:"repo"`
+	Type    string `yaml:"type"`
+}
+
+type pnpmLockPackage struct {
+	Resolution pnpmLockPackageResolution `yaml:"resolution"`
+	Name       string                    `yaml:"name"`
+	Version    string                    `yaml:"version"`
+	Dev        bool                      `yaml:"dev"`
+}
+
+type pnpmLockfile struct {
+	Version  float64                    `yaml:"lockfileVersion"`
+	Packages map[string]pnpmLockPackage `yaml:"packages,omitempty"`
+}
+
+type pnpmLockfileV6 struct {
+	Version  string                     `yaml:"lockfileVersion"`
+	Packages map[string]pnpmLockPackage `yaml:"packages,omitempty"`
+}
+
+// UnmarshalYAML is a custom unmarshaling function for handling v6 lockfiles.
+func (l *pnpmLockfile) UnmarshalYAML(unmarshal func(any) error) error {
+	var lockfileV6 pnpmLockfileV6
+
+	if err := unmarshal(&lockfileV6); err != nil {
+		return err
+	}
+
+	parsedVersion, err := strconv.ParseFloat(lockfileV6.Version, 64)
+
+	if err != nil {
+		return err
+	}
+
+	l.Version = parsedVersion
+	l.Packages = lockfileV6.Packages
+
+	return nil
+}
+
+// PnpmEcosystem is the OSV ecpsystem for pnpm-lock.yaml files.
+const PnpmEcosystem = "npm"
+
+var (
+	numberMatcher = regexp.MustCompile(`^\d`)
+	// Looks for the pattern "name@version", where name is allowed to contain zero or more "@"
+	nameVersionRegexp = regexp.MustCompile(`^(.+)@([\d.]+)$`)
+)
+
+func startsWithNumber(str string) bool {
+	return numberMatcher.MatchString(str)
+}
+
+// extractPnpmPackageNameAndVersion parses a dependency path, attempting to
+// extract the name and version of the package it represents
+func extractPnpmPackageNameAndVersion(dependencyPath string, lockfileVersion float64) (string, string) {
+	// file dependencies must always have a name property to be installed,
+	// and their dependency path never has the version encoded, so we can
+	// skip trying to extract either from their dependency path
+	if strings.HasPrefix(dependencyPath, "file:") {
+		return "", ""
+	}
+
+	// v9.0 specifies the dependencies as <package>@<version> rather than as a path
+	if lockfileVersion == 9.0 {
```

Comment:
This should be >= to account for future versions. I believe pnpm now is pinning the lockfile version to their release version, so even if the lockfile format doesn't change, the lockfile version will increase.